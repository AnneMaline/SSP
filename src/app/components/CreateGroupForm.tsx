"use client";
import { useState } from "react";

type CreateGroupFormType = {
  environment: "prod" | "test" | "development";
  type: "data" | "service" | "users";
  name: string;
  accessType: "view" | "data" | "edit";
};

const CreateGroupForm = () => {
  // Initial state for the form
  const initialFormData: CreateGroupFormType = {
    environment: "prod",
    type: "data",
    name: "",
    accessType: "view",
  };
  const environments: CreateGroupFormType["environment"][] = [
    "prod",
    "test",
    "development",
  ];
  const types: CreateGroupFormType["type"][] = ["data", "service", "users"];
  const accessTypes: CreateGroupFormType["accessType"][] = [
    "view",
    "data",
    "edit",
  ];

  // feedback and questions form
  const [formData, setFormData] =
    useState<CreateGroupFormType>(initialFormData);

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    console.log("Submitting form data:", formData);

    // Reset form after submission
    setFormData(initialFormData);

    /*// Send data to the server or API
    const response = await fetch("/api/submit-form", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(formData),
    });

    if (response.ok) {
      console.log("Form submitted successfully!");
    } else {
      console.error("Form submission failed.");
    }*/
  };
  return (
    <form
      onSubmit={handleSubmit}
      className="max-w-md mx-auto p-6 border border-gray-300 rounded-lg shadow space-y-6"
    >
      {/* Environment */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Environment</legend>
        <div className="space-y-1">
          {environments.map((type) => (
            <label key={type} className="block">
              <input
                type="radio"
                name="environment"
                value={type}
                checked={formData.environment === type}
                onChange={() =>
                  setFormData((prev) => ({
                    ...prev,
                    environment: type as CreateGroupFormType["environment"],
                  }))
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Group Type */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Group Type</legend>
        <div className="space-y-1">
          {types.map((type) => (
            <label key={type} className="block">
              <input
                type="radio"
                name="type"
                value={type}
                checked={formData.type === type}
                onChange={() =>
                  setFormData((prev) => ({
                    ...prev,
                    type: type as CreateGroupFormType["type"],
                  }))
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Group Name */}
      <div className="space-y-2">
        <label htmlFor="name" className="block text-lg font-semibold">
          Group Name
        </label>
        <input
          required
          type="text"
          id="name"
          name="name"
          value={formData.name}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setFormData((prev) => ({
              ...prev,
              name: e.target.value,
            }))
          }
          className="w-full px-3 py-2 border border-gray-300 rounded"
          placeholder="Enter group name"
        />
      </div>

      {/* Access Type */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Access Type</legend>
        <div className="space-y-1">
          {accessTypes.map((type) => (
            <label key={type} className="block">
              <input
                type="radio"
                name="accessType"
                value={type}
                checked={formData.accessType === type}
                onChange={() =>
                  setFormData((prev) => ({
                    ...prev,
                    accessType: type as CreateGroupFormType["accessType"],
                  }))
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Submit Button */}
      <button
        type="submit"
        className="w-full bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-600"
      >
        Create
      </button>
    </form>
  );
};

export default CreateGroupForm;
