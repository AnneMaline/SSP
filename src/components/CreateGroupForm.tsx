"use client";
import { useState } from "react";
import { createGroup } from "@/utils/entitlement/createGroup";

type CreateGroupFormType = {
  environment: "prod" | "test" | "development" | "bootcamp";
  type: "data" | "service" | "users";
  name: string;
  description: string;
  accessType: "view" | "data" | "edit";
};

const CreateGroupForm = () => {
  // Initial state for the form
  const initialFormData: CreateGroupFormType = {
    environment: "bootcamp",
    type: "data",
    name: "string",
    description: "string",
    accessType: "view",
  };
  const environments: CreateGroupFormType["environment"][] = [
    "prod",
    "test",
    "development",
    "bootcamp",
  ];
  const types: CreateGroupFormType["type"][] = ["data", "service", "users"];
  const accessTypes: CreateGroupFormType["accessType"][] = [
    "view",
    "data",
    "edit",
  ];

  // feedback and questions form
  const [formData, setFormData] =
    useState<CreateGroupFormType>(initialFormData);

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    // prevent the default form submission
    e.preventDefault();

    // Reset form after submission
    setFormData(initialFormData);

    // Send data to the API
    createGroup(formData.name, formData.description, "bootcamp");
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="max-w-md mx-auto p-6 border-2 border-dashed border-[#80AFB4] rounded-lg  space-y-6"
    >
      {/* Environment */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Environment</legend>
        <div className="space-y-1">
          {environments.map((type) => (
            <label key={type} className="block">
              <input
                type="radio"
                name="environment"
                value={type}
                checked={formData.environment === type}
                onChange={() =>
                  setFormData((prev) => ({
                    ...prev,
                    environment: type as CreateGroupFormType["environment"],
                  }))
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Group Type */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Group Type</legend>
        <div className="space-y-1">
          {types.map((type) => (
            <label key={type} className="block">
              <input
                type="radio"
                name="type"
                value={type}
                checked={formData.type === type}
                onChange={() =>
                  setFormData((prev) => ({
                    ...prev,
                    type: type as CreateGroupFormType["type"],
                  }))
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Group Name */}
      <div className="space-y-2">
        <label htmlFor="name" className="block text-lg font-semibold">
          Group Name
        </label>
        <input
          required
          type="text"
          id="name"
          name="name"
          value={formData.name}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setFormData((prev) => ({
              ...prev,
              name: e.target.value,
            }))
          }
          className="w-full px-3 py-2 border border-gray-300 rounded"
          placeholder="Enter group name"
        />
      </div>

      {/* Access Type */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Access Type</legend>
        <div className="space-y-1">
          {accessTypes.map((type) => (
            <label key={type} className="block">
              <input
                type="radio"
                name="accessType"
                value={type}
                checked={formData.accessType === type}
                onChange={() =>
                  setFormData((prev) => ({
                    ...prev,
                    accessType: type as CreateGroupFormType["accessType"],
                  }))
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Submit Button */}
      <button
        type="submit"
        className="w-full bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-600"
      >
        Create
      </button>
    </form>
  );
};

export default CreateGroupForm;
