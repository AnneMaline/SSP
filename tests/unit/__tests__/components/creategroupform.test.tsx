import { render, screen, fireEvent, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import CreateGroupForm from "../../../../src/components/CreateGroupForm";

describe("CreateGroupForm Component", () => {
  it("renders correctly", () => {
    const { container } = render(<CreateGroupForm />);
    expect(container).toMatchSnapshot();
  });

  it("renders the form with all fields and options", () => {
    render(<CreateGroupForm />);

    // Check environment radios
    expect(screen.getByLabelText("Prod")).toBeInTheDocument();
    expect(screen.getByLabelText("Test")).toBeInTheDocument();
    expect(screen.getByLabelText("Development")).toBeInTheDocument();

    // Check group type radios
    const groupTypeFieldset = screen
      .getByText("Group Type")
      .closest("fieldset");
    expect(groupTypeFieldset).toBeInTheDocument();
    const groupTypeRadioGroup = within(groupTypeFieldset!);
    expect(groupTypeRadioGroup.getByLabelText("Data")).toBeInTheDocument();
    expect(groupTypeRadioGroup.getByLabelText("Service")).toBeInTheDocument();
    expect(groupTypeRadioGroup.getByLabelText("Users")).toBeInTheDocument();

    // Check access type radios
    const accessTypeFieldset = screen
      .getByText("Access Type")
      .closest("fieldset");
    expect(accessTypeFieldset).toBeInTheDocument();
    const accessTypeRadioGroup = within(accessTypeFieldset!);
    expect(accessTypeRadioGroup.getByLabelText("View")).toBeInTheDocument();
    expect(accessTypeRadioGroup.getByLabelText("Data")).toBeInTheDocument();
    expect(accessTypeRadioGroup.getByLabelText("Edit")).toBeInTheDocument();

    // Check name input
    expect(screen.getByPlaceholderText("Enter group name")).toBeInTheDocument();

    // Check submit button
    expect(screen.getByRole("button", { name: "Create" })).toBeInTheDocument();
  });

  it("updates the environment value when selected", () => {
    render(<CreateGroupForm />);
    const prodRadio = screen.getByLabelText("Prod");
    const testRadio = screen.getByLabelText("Test");

    // Check initial state
    expect(prodRadio).toBeChecked();
    expect(testRadio).not.toBeChecked();

    // Select "Test"
    fireEvent.click(testRadio);

    // Verify state update
    expect(prodRadio).not.toBeChecked();
    expect(testRadio).toBeChecked();
  });

  it("updates the group type value when selected", () => {
    render(<CreateGroupForm />);

    const groupTypeFieldset = screen
      .getByText("Group Type")
      .closest("fieldset");
    const groupTypeRadioGroup = within(groupTypeFieldset!);

    const dataRadio = groupTypeRadioGroup.getByLabelText("Data");
    const serviceRadio = groupTypeRadioGroup.getByLabelText("Service");

    // Check initial state
    expect(dataRadio).toBeChecked();
    expect(serviceRadio).not.toBeChecked();

    // Select "Service"
    fireEvent.click(serviceRadio);

    // Verify state update
    expect(dataRadio).not.toBeChecked();
    expect(serviceRadio).toBeChecked();
  });

  it("updates the access type value when selected", () => {
    render(<CreateGroupForm />);

    const accessTypeFieldset = screen
      .getByText("Access Type")
      .closest("fieldset");
    const accessTypeRadioGroup = within(accessTypeFieldset!);

    const viewRadio = accessTypeRadioGroup.getByLabelText("View");
    const editRadio = accessTypeRadioGroup.getByLabelText("Edit");

    // Check initial state
    expect(viewRadio).toBeChecked();
    expect(editRadio).not.toBeChecked();

    // Select "Edit"
    fireEvent.click(editRadio);

    // Verify state update
    expect(viewRadio).not.toBeChecked();
    expect(editRadio).toBeChecked();
  });

  it("updates the group name value when typed", () => {
    render(<CreateGroupForm />);
    const nameInput = screen.getByPlaceholderText("Enter group name");

    // Check initial state
    expect(nameInput).toHaveValue("");

    // Type into the input
    fireEvent.change(nameInput, { target: { value: "Test Group" } });

    // Verify state update
    expect(nameInput).toHaveValue("Test Group");
  });

  it("submits the form and resets the state", () => {
    render(<CreateGroupForm />);
    const nameInput = screen.getByPlaceholderText("Enter group name");
    const submitButton = screen.getByRole("button", { name: "Create" });

    // Fill out the form
    fireEvent.change(nameInput, { target: { value: "Test Group" } });
    const testRadio = screen.getByLabelText("Test");
    fireEvent.click(testRadio);

    const accessTypeFieldset = screen
      .getByText("Access Type")
      .closest("fieldset");
    const accessTypeRadioGroup = within(accessTypeFieldset!);
    fireEvent.click(accessTypeRadioGroup.getByLabelText("Edit"));

    const groupTypeFieldset = screen
      .getByText("Group Type")
      .closest("fieldset");
    const groupTypeRadioGroup = within(groupTypeFieldset!);
    fireEvent.click(groupTypeRadioGroup.getByLabelText("Service"));

    // Submit the form
    fireEvent.click(submitButton);

    // Verify the state has reset
    expect(nameInput).toHaveValue("");
    expect(screen.getByLabelText("Prod")).toBeChecked(); // Default environment
    expect(groupTypeRadioGroup.getByLabelText("Data")).toBeChecked(); // Default group type
    expect(accessTypeRadioGroup.getByLabelText("View")).toBeChecked(); // Default access type
  });
});
