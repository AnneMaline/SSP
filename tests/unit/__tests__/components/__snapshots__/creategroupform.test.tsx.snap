// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateGroupForm Component renders correctly 1`] = `
<div>
  <form
    class="max-w-md mx-auto p-6 border-2 border-dashed border-[#80AFB4] rounded-lg  space-y-6"
  >
    <fieldset
      class="space-y-2"
    >
      <legend
        class="block text-lg font-semibold"
      >
        Environment
      </legend>
      <div
        class="space-y-1"
      >
        <label
          class="block"
        >
          <input
            checked=""
            class="mr-2"
            name="environment"
            type="radio"
            value="prod"
          />
          Prod
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="environment"
            type="radio"
            value="test"
          />
          Test
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="environment"
            type="radio"
            value="development"
          />
          Development
        </label>
      </div>
    </fieldset>
    <fieldset
      class="space-y-2"
    >
      <legend
        class="block text-lg font-semibold"
      >
        Group Type
      </legend>
      <div
        class="space-y-1"
      >
        <label
          class="block"
        >
          <input
            checked=""
            class="mr-2"
            name="type"
            type="radio"
            value="data"
          />
          Data
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="type"
            type="radio"
            value="service"
          />
          Service
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="type"
            type="radio"
            value="users"
          />
          Users
        </label>
      </div>
    </fieldset>
    <div
      class="space-y-2"
    >
      <label
        class="block text-lg font-semibold"
        for="name"
      >
        Group Name
      </label>
      <input
        class="w-full px-3 py-2 border border-gray-300 rounded"
        id="name"
        name="name"
        placeholder="Enter group name"
        required=""
        type="text"
        value=""
      />
    </div>
    <fieldset
      class="space-y-2"
    >
      <legend
        class="block text-lg font-semibold"
      >
        Access Type
      </legend>
      <div
        class="space-y-1"
      >
        <label
          class="block"
        >
          <input
            checked=""
            class="mr-2"
            name="accessType"
            type="radio"
            value="view"
          />
          View
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="accessType"
            type="radio"
            value="data"
          />
          Data
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="accessType"
            type="radio"
            value="edit"
          />
          Edit
        </label>
      </div>
    </fieldset>
    <button
      class="w-full bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-600"
      type="submit"
    >
      Create
    </button>
  </form>
</div>
`;
